pytest fails to find/bind fixture: "AttributeError: skipif_no_docker_engine"

Things were working and then they stopped... tried in another clone with the same 'success'. 
 anyone has a clue?

Here is a full protocol, hopefully without leaking any credentials ;)

```shell
(git)hopa:~/proj/repronim/reproman-master[master]git
$> python -m pytest -s -v reproman                                                                   
===================================================== test session starts =====================================================
platform linux -- Python 3.7.2, pytest-3.10.1, py-1.7.0, pluggy-0.8.0 -- /home/yoh/proj/repronim/reproman-master/venvs/dev3/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/yoh/proj/repronim/reproman-master/.hypothesis/examples')
rootdir: /home/yoh/proj/repronim/reproman-master, inifile:
plugins: localserver-0.5.0, hypothesis-3.71.11
collecting 135 items / 2 errors                                                                                               2019-05-30 12:06:28,320 [ERROR  ] stderr| Error: communication error
| CEDAR:6001:Failed to connect to <127.0.0.1:9618> 
2019-05-30 12:06:28,320 [ERROR  ] Failed to run ['condor_status'] under None. Exit code=1. out= err=Error: communication error
| CEDAR:6001:Failed to connect to <127.0.0.1:9618>
|  
collected 374 items / 2 errors                                                                                                

=========================================================== ERRORS ============================================================
________________________________ ERROR collecting reproman/distributions/tests/test_docker.py _________________________________
reproman/distributions/tests/test_docker.py:23: in <module>
    @mark.skipif_no_docker_engine
reproman/tests/skip.py:207: in __getattr__
    return super(Mark, self).__getattr__(item)
reproman/tests/skip.py:166: in __getattr__
    raise AttributeError(item) from None
E   AttributeError: skipif_no_docker_engine
______________________________ ERROR collecting reproman/resource/tests/test_docker_container.py ______________________________
reproman/resource/tests/test_docker_container.py:168: in <module>
    @mark.skipif_no_docker_engine
reproman/tests/skip.py:207: in __getattr__
    return super(Mark, self).__getattr__(item)
reproman/tests/skip.py:166: in __getattr__
    raise AttributeError(item) from None
E   AttributeError: skipif_no_docker_engine
===============================================
```

<details>
<summary>and even fullter</summary>

```
(git)hopa:~/proj/repronim/reproman-master[master]git
$> virtualenv --system-site-packages --python python3 venvs/dev3
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/yoh/proj/repronim/reproman-master/venvs/dev3/bin/python3
Also creating executable in /home/yoh/proj/repronim/reproman-master/venvs/dev3/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
1 12721.....................................:Thu 30 May 2019 12:07:07 PM EDT:.
(git)hopa:~/proj/repronim/reproman-master[master]git
$> source venvs/dev3/bin/activate                               
(dev3) 1 12723.....................................:Thu 30 May 2019 12:07:44 PM EDT:.
(git)hopa:~/proj/repronim/reproman-master[master]git
$> pip install -e .
Obtaining file:///home/yoh/proj/repronim/reproman-master
Requirement already satisfied: appdirs in /usr/lib/python3/dist-packages (from reproman==0.1.0) (1.4.3)
Requirement already satisfied: attrs>=16.3.0 in /usr/lib/python3/dist-packages (from reproman==0.1.0) (18.2.0)
Requirement already satisfied: humanize in /usr/lib/python3/dist-packages (from reproman==0.1.0) (0.5.1)
Requirement already satisfied: pyyaml in /usr/lib/python3/dist-packages (from reproman==0.1.0) (3.13)
Requirement already satisfied: tqdm in /usr/lib/python3/dist-packages (from reproman==0.1.0) (4.28.1)
Collecting fabric>=2.3.1 (from reproman==0.1.0)
  Using cached https://files.pythonhosted.org/packages/d9/e4/e6fa248c94ee5d45def54b609fcf70f39d0b7f7050f2d4405c5f156b5516/fabric-2.4.0-py2.py3-none-any.whl
Requirement already satisfied: cryptography>=1.5 in /usr/lib/python3/dist-packages (from reproman==0.1.0) (2.6.1)
Requirement already satisfied: pytz in /usr/lib/python3/dist-packages (from reproman==0.1.0) (2018.9)
Requirement already satisfied: scp in /usr/lib/python3/dist-packages (from reproman==0.1.0) (0.13.0)
Requirement already satisfied: pycrypto in /usr/lib/python3/dist-packages (from reproman==0.1.0) (2.6.1)
Collecting pyOpenSSL==16.2.0 (from reproman==0.1.0)
  Using cached https://files.pythonhosted.org/packages/ac/93/b4cd538d31adacd07f83013860db6b88d78755af1f3fefe68ec22d397e7b/pyOpenSSL-16.2.0-py2.py3-none-any.whl
Requirement already satisfied: requests in /usr/lib/python3/dist-packages (from reproman==0.1.0) (2.21.0)
Requirement already satisfied: rpaths in /usr/lib/python3/dist-packages (from reproman==0.1.0) (0.13)
Requirement already satisfied: jinja2 in /usr/lib/python3/dist-packages (from reproman==0.1.0) (2.10)
Requirement already satisfied: reprozip in /usr/lib/python3/dist-packages (from reproman==0.1.0) (1.0.14)
Collecting invoke<2.0,>=1.1 (from fabric>=2.3.1->reproman==0.1.0)
  Using cached https://files.pythonhosted.org/packages/be/9f/8508712c9cad73ac0c8eeb2c3e51c9ef65136653dda2b512bde64109f023/invoke-1.2.0-py3-none-any.whl
Requirement already satisfied: paramiko>=2.4 in /usr/lib/python3/dist-packages (from fabric>=2.3.1->reproman==0.1.0) (2.4.2)
Requirement already satisfied: six>=1.5.2 in /usr/lib/python3/dist-packages (from pyOpenSSL==16.2.0->reproman==0.1.0) (1.12.0)
Installing collected packages: invoke, fabric, pyOpenSSL, reproman
  Found existing installation: pyOpenSSL 19.0.0
    Not uninstalling pyopenssl at /usr/lib/python3/dist-packages, outside environment /home/yoh/proj/repronim/reproman-master/venvs/dev3
    Can't uninstall 'pyOpenSSL'. No files were found to uninstall.
  Running setup.py develop for reproman
Successfully installed fabric-2.4.0 invoke-1.2.0 pyOpenSSL-16.2.0 reproman
(dev3) 1 12724.....................................:Thu 30 May 2019 12:07:52 PM EDT:.
(git)hopa:~/proj/repronim/reproman-master[master]git
$> python -m pytest -s -v reproman                              
===================================================== test session starts =====================================================
platform linux -- Python 3.7.3rc1, pytest-3.10.1, py-1.7.0, pluggy-0.8.0 -- /home/yoh/proj/repronim/reproman-master/venvs/dev3/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/yoh/proj/repronim/reproman-master/.hypothesis/examples')
rootdir: /home/yoh/proj/repronim/reproman-master, inifile:
plugins: localserver-0.5.0, hypothesis-3.71.11
collecting 59 items / 1 errors                                                                                                2019-05-30 12:07:57,609 [ERROR  ] stderr| Error: communication error
| CEDAR:6001:Failed to connect to <127.0.0.1:9618> 
2019-05-30 12:07:57,609 [ERROR  ] Failed to run ['condor_status'] under None. Exit code=1. out= err=Error: communication error
| CEDAR:6001:Failed to connect to <127.0.0.1:9618>
|  
collected 374 items / 2 errors                                                                                                

=========================================================== ERRORS ============================================================
________________________________ ERROR collecting reproman/distributions/tests/test_docker.py _________________________________
reproman/distributions/tests/test_docker.py:23: in <module>
    @mark.skipif_no_docker_engine
reproman/tests/skip.py:207: in __getattr__
    return super(Mark, self).__getattr__(item)
reproman/tests/skip.py:166: in __getattr__
    raise AttributeError(item) from None
E   AttributeError: skipif_no_docker_engine
______________________________ ERROR collecting reproman/resource/tests/test_docker_container.py ______________________________
reproman/resource/tests/test_docker_container.py:168: in <module>
    @mark.skipif_no_docker_engine
reproman/tests/skip.py:207: in __getattr__
    return super(Mark, self).__getattr__(item)
reproman/tests/skip.py:166: in __getattr__
    raise AttributeError(item) from None
E   AttributeError: skipif_no_docker_engine
====================================================== warnings summary =======================================================
/usr/lib/python3/dist-packages/pyasn1/type/constraint.py:130
  /usr/lib/python3/dist-packages/pyasn1/type/constraint.py:130: DeprecationWarning: invalid escape sequence \*
    """

/usr/lib/python3/dist-packages/pyasn1/type/constraint.py:175
  /usr/lib/python3/dist-packages/pyasn1/type/constraint.py:175: DeprecationWarning: invalid escape sequence \*
    """

/usr/lib/python3/dist-packages/pyasn1/type/constraint.py:334
  /usr/lib/python3/dist-packages/pyasn1/type/constraint.py:334: DeprecationWarning: invalid escape sequence \*
    """

/usr/lib/python3/dist-packages/pyasn1/type/constraint.py:494
  /usr/lib/python3/dist-packages/pyasn1/type/constraint.py:494: DeprecationWarning: invalid escape sequence \*
    """

/usr/lib/python3/dist-packages/pyasn1/type/constraint.py:538
  /usr/lib/python3/dist-packages/pyasn1/type/constraint.py:538: DeprecationWarning: invalid escape sequence \*
    """

/usr/lib/python3/dist-packages/pyasn1/type/namedval.py:54
  /usr/lib/python3/dist-packages/pyasn1/type/namedval.py:54: DeprecationWarning: invalid escape sequence \*
    """

/usr/lib/python3/dist-packages/datalad/utils.py:75
  /usr/lib/python3/dist-packages/datalad/utils.py:75: DeprecationWarning: dist() and linux_distribution() functions are deprecated in Python 3.5
    = platform.linux_distribution()[:2]

/usr/lib/python3/dist-packages/datalad/utils.py:961
  /usr/lib/python3/dist-packages/datalad/utils.py:961: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    is_decorating = not kwargs and len(args) == 1 and isinstance(args[0], collections.Callable)

/usr/lib/python3/dist-packages/html5lib/_trie/_base.py:3
  /usr/lib/python3/dist-packages/html5lib/_trie/_base.py:3: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Mapping

-- Docs: https://docs.pytest.org/en/latest/warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================= 9 warnings, 2 error in 1.46 seconds ```
</details>

<details><summary>just another proof that I am not insane!</summary>

```shell
$> git clone http://github.com/ReproNim/reproman && cd reproman && virtualenv --system-site-packages --python python3 venvs/dev3 && source venvs/dev3/bin/activate && python -m pytest -s -v reproman 
Cloning into 'reproman'...                                             
warning: redirecting to https://github.com/ReproNim/reproman/
remote: Enumerating objects: 69, done.
remote: Counting objects: 100% (69/69), done.
remote: Compressing objects: 100% (42/42), done.
remote: Total 11839 (delta 32), reused 42 (delta 27), pack-reused 11770
Receiving objects: 100% (11839/11839), 2.70 MiB | 1.79 MiB/s, done.
Resolving deltas: 100% (9170/9170), done.
CHANGELOG.md     Dockerfile-debian  Singularity         formatters.py    reproman.cfg            setup.cfg         tools/
CONTRIBUTING.md  Dockerfile-redhat  docker-compose.yml  readthedocs.yml  requirements-devel.txt  setup.py*         tox.ini
COPYING          README.md          docs/               reproman/        requirements.txt        setup_support.py
Running virtualenv with interpreter /home/yoh/proj/repronim/reproman-master/venvs/dev3/bin/python3
Using real prefix '/usr'
Path not in prefix '/home/yoh/proj/repronim/reproman-master/venvs/dev3/include/python3.7m' '/usr'
New python executable in /tmp/reproman/venvs/dev3/bin/python3
Also creating executable in /tmp/reproman/venvs/dev3/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
===================================================== test session starts =====================================================
platform linux -- Python 3.7.3rc1, pytest-3.10.1, py-1.7.0, pluggy-0.8.0 -- /tmp/reproman/venvs/dev3/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/tmp/reproman/.hypothesis/examples')
rootdir: /tmp/reproman, inifile:
plugins: localserver-0.5.0, hypothesis-3.71.11
collecting 124 items / 1 errors                                                                                               2019-05-30 15:32:11,786 [ERROR  ] stderr| Error: communication error
| CEDAR:6001:Failed to connect to <127.0.0.1:9618> 
2019-05-30 15:32:11,786 [ERROR  ] Failed to run ['condor_status'] under None. Exit code=1. out= err=Error: communication error
| CEDAR:6001:Failed to connect to <127.0.0.1:9618>
|  
collected 355 items / 4 errors                                                                                                

=========================================================== ERRORS ============================================================
________________________________ ERROR collecting reproman/distributions/tests/test_docker.py _________________________________
reproman/distributions/tests/test_docker.py:23: in <module>
    @mark.skipif_no_docker_engine
reproman/tests/skip.py:207: in __getattr__
    return super(Mark, self).__getattr__(item)
reproman/tests/skip.py:166: in __getattr__
    raise AttributeError(item) from None
E   AttributeError: skipif_no_docker_engine
______________________________ ERROR collecting reproman/resource/tests/test_docker_container.py ______________________________
reproman/resource/tests/test_docker_container.py:168: in <module>
    @mark.skipif_no_docker_engine
reproman/tests/skip.py:207: in __getattr__
    return super(Mark, self).__getattr__(item)
reproman/tests/skip.py:166: in __getattr__
    raise AttributeError(item) from None
E   AttributeError: skipif_no_docker_engine
__________________________________ ERROR collecting reproman/resource/tests/test_session.py ___________________________________
ImportError while importing test module '/tmp/reproman/reproman/resource/tests/test_session.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
reproman/resource/tests/test_session.py:12: in <module>
    from fabric import Connection
E   ModuleNotFoundError: No module named 'fabric'
___________________________________ ERROR collecting reproman/resource/tests/test_shell.py ____________________________________
ImportError while importing test module '/tmp/reproman/reproman/resource/tests/test_shell.
```
</details>

ok -- the actual reason was masked away by
```
    def __getattr__(self, item):
        if item.startswith("skipif_"):
            try:
                return super(Mark, self).__getattr__(item[len("skipif_"):])
            except AttributeError:
                # Fall back to the original item name so that the attribute
                # error message doesn't confusingly drop "skipif_".
                pass
        return super(Mark, self).__getattr__(item)
```
where my system wide docker-py was probably too old or new (3.4.1-4) and the check above was throwing up with 
```
______________________ ERROR collecting reproman/resource/tests/test_docker_container.py _______________________
reproman/tests/skip.py:202: in __getattr__
    return super(Mark, self).__getattr__(item[len("skipif_"):])
reproman/tests/skip.py:167: in __getattr__
    return self.attr_value(condfn)
reproman/tests/skip.py:196: in attr_value
    reason, cond = condition_func()
reproman/tests/skip.py:103: in no_docker_engine
    return "docker engine not running", not is_engine_running()
reproman/tests/skip.py:97: in is_engine_running
    return DockerContainer.is_engine_running()
reproman/resource/docker_container.py:74: in is_engine_running
    session = docker.Client(base_url=base_url)
E   AttributeError: module 'docker' has no attribute 'Client'
```
which was happily caught in above code and "pass"ed on.  IMHO

1. this try/except should be avoided if possible or somehow match only specific exception instances
2. check for having compatible `docker` module be fortified

--- a/reproman/tests/skip.py
+++ b/reproman/tests/skip.py
@@ -148,6 +148,15 @@ CONDITION_FNS = [
 # Entry points: skipif and mark
 
 
+class NamespaceAttributeError(AttributeError):
+    """Namespace-specific AttributeError.
+
+    Raised by Namespace when it cannot find the specified condition function.
+    Using a derived class allows us to distinguish an unknown condition
+    function from a condition function that raises an AttributeError.
+    """
+
+
 class Namespace(object, metaclass=abc.ABCMeta):
     """Provide namespace skip conditions in CONDITION_FNS.
     """
@@ -163,7 +172,7 @@ class Namespace(object, metaclass=abc.ABCMeta):
         try:
             condfn = self.fns[item]
         except KeyError:
-            raise AttributeError(item) from None
+            raise NamespaceAttributeError(item) from None
         return self.attr_value(condfn)
 
 
@@ -200,7 +209,7 @@ class Mark(Namespace):
         if item.startswith("skipif_"):
             try:
                 return super(Mark, self).__getattr__(item[len("skipif_"):])
-            except AttributeError:
+            except NamespaceAttributeError:
                 # Fall back to the original item name so that the attribute
                 # error message doesn't confusingly drop "skipif_".
                 pass

--- a/reproman/tests/test_skip.py
+++ b/reproman/tests/test_skip.py
@@ -10,6 +10,7 @@ from unittest.mock import patch
 
 import pytest
 
+from reproman.tests.skip import Mark
 from reproman.tests.skip import mark
 from reproman.tests.skip import skipif
 
@@ -47,3 +48,18 @@ def test_skipif_unknown_attribute():
 def test_mark_skipif_unknown_attribute():
     with pytest.raises(AttributeError):
         mark.skipif_youdontknowme
+
+
+def test_other_attribute_error():
+    # If a condition function raises an AttributeError, Mark doesn't mistake it
+    # for an unknown condition function.
+
+    def no_confusion():
+        raise AttributeError("don't get confused")
+
+    with patch("reproman.tests.skip.Namespace.fns",
+               {"no_confusion": no_confusion}):
+        with pytest.raises(AttributeError) as exc:
+            m = Mark()
+            m.skipif_no_confusion
+        assert "don't get confused" in str(exc)

